cmd/console-proxy/src/posix.rs

```diff
@@ -0,0 +1,401 @@
+        )?;
+
+        let v = match value {
+            Ok(v) => v,
+            Err(e) => bail!("Got Hiffy error: {e}"),
+        };
+
+        let v =
+            v.as_base()?.as_u32().ok_or_else(|| anyhow!("Couldn't get U32"))?;
+
+        Ok(v as usize)
+    }
+
+    fn write(&mut self, buf: &[u8]) -> Result<usize> {
+        let op = IdolOperation::new(
+            self.hubris,
+            "ControlPlaneAgent",
+            "uart_write",
+            None,
+        )
+        .context("Could not find `ControlPlaneAgent.uart_write` operation")?;
+
+        let buf = &buf[..usize::min(buf.len(), HIFFY_BUF_SIZE)];
+
+        let value = humility_cmd_hiffy::hiffy_call(
+            self.hubris,
+            self.core,
+            &mut self.context,
+            &op,
+            &[],
+            Some(HiffyLease::Write(buf)),
+        )?;
+
+        let v = match value {
+            Ok(v) => v,
+            Err(e) => bail!("Got Hiffy error: {e}"),
+        };
+
+        let v =
+            v.as_base()?.as_u32().ok_or_else(|| anyhow!("Couldn't get U32"))?;
+
+        Ok(v as usize)
+    }
+
+    fn attach(
+        &mut self,
+        raw: bool,
+        imap: RemapRules,
+        omap: RemapRules,
+        log: Option<File>,
+    ) -> Result<()> {
+        // Put terminal in raw mode, if requested, with a guard to restore it.
+        let _guard = if raw {
+            Some(UnrawTermiosGuard::make_stdout_raw()?)
+        } else {
+            None
+        };
+
+        let (stdin_tx, stdin_rx) = crossbeam_channel::unbounded();
+        let (stdout_tx, stdout_rx) = crossbeam_channel::unbounded();
+        thread::spawn(move || stdin_reader(raw, stdin_tx, omap));
+        thread::spawn(move || stdout_writer(stdout_rx, imap, log));
+
+        let mut rx_buf = vec![0; HIFFY_BUF_SIZE];
+        let mut tx_buf = Vec::new();
+        loop {
+            if !tx_buf.is_empty() {
+                let nwritten = self.write(&tx_buf)?;
+                tx_buf.drain(0..nwritten);
+            }
+
+            let nread = self.read(&mut rx_buf)?;
+            if nread > 0 {
+                stdout_tx.send(rx_buf[..nread].to_owned()).unwrap();
+            }
+
+            // If our read returned the max buffer size, don't wait for
+            // `poll_interval` to try again, just wait 1ms to give the other
+            // channels a chance to sneak something in.
+            let timeout = if nread == HIFFY_BUF_SIZE {
+                Duration::from_millis(1)
+            } else {
+                self.poll_interval
+            };
+            let timeout = crossbeam_channel::after(timeout);
+
+            let mut done = false;
+            loop {
+                select! {
+                    recv(stdin_rx) -> data => {
+                        if let Ok(mut data) = data {
+                            tx_buf.append(&mut data);
+                        } else {
+                            done = true;
+                            break;
```

Comment:
This could be done with a loop label instead of a flag, e.g. `'outer: loop` and `break 'outer;`